lib/Service/ReverseGeoCoderService.php

```diff
@@ -0,0 +1,163 @@
+<?php
+
+declare(strict_types=1);
+/**
+ * @copyright Copyright (c) 2022 Louis Chemineau <louis@chmn.me>
+ *
+ * @author Louis Chemineau <louis@chmn.me>
+ *
+ * @license AGPL-3.0-or-later
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+namespace OCA\Photos\Service;
+
+use OCP\Files\IAppData;
+use OCP\Files\SimpleFS\ISimpleFolder;
+use OCP\Files\NotFoundException;
+use OCP\Http\Client\IClientService;
+use Hexogen\KDTree\FSTreePersister;
+use Hexogen\KDTree\FSKDTree;
+use Hexogen\KDTree\KDTree;
+use Hexogen\KDTree\Item;
+use Hexogen\KDTree\ItemList;
+use Hexogen\KDTree\ItemFactory;
+use Hexogen\KDTree\NearestSearch;
+use Hexogen\KDTree\Point;
+
+class ReverseGeoCoderService {
+	private ISimpleFolder $geoNameFolder;
+	private ?NearestSearch $fsSearcher = null;
+	/** @var array<int, string> */
+	private ?array $citiesMapping = null;
+
+	public function __construct(
+		IAppData $appData,
+		private IClientService $clientService,
+	) {
+		try {
+			$this->geoNameFolder = $appData->getFolder("geonames");
+		} catch (NotFoundException $ex) {
+			$this->geoNameFolder = $appData->newFolder("geonames");
+		}
+	}
+
+	public function getLocationForCoordinates(float $latitude, float $longitude): string {
+		if ($this->fsSearcher === null) {
+			$this->buildKDTree();
+			$kdTreeFileContent = $this->geoNameFolder->getFile("cities1000.bin")->getContent();
+			$kdTreeTmpFileName = tempnam(sys_get_temp_dir(), "nextcloud_photos_");
+			file_put_contents($kdTreeTmpFileName, $kdTreeFileContent);
+			$fsTree = new FSKDTree($kdTreeTmpFileName, new ItemFactory());
+			$this->fsSearcher = new NearestSearch($fsTree);
+		}
+
+		$result = $this->fsSearcher->search(new Point([$latitude, $longitude]), 1);
+		return $this->getLocationNameForLocationId($result[0]->getId());
+	}
+
+	private function getLocationNameForLocationId(int $locationId): string {
```

Comment:
Only loaded during the mapping, which is done in a background job, so it does not blow up on every request.